bpo-37587: Make json.loads faster for long strings

mpaolini opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.
Make json.loads faster for long strings
# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-37587](https://bugs.python.org/issue37587) -->
https://bugs.python.org/issue37587
<!-- /issue-number -->

mpaolini committed:
bpo-37587: Make json.loads faster for long strings

Forces the compiler to use a register variable for a tight loop
in the hot-path.

It also optimizes a condition for the common case strict=true.

serhiy-storchaka reviewed (commented):
`c` is only used in the condition `if (!(c == '"' || c == '\\'))` after the loop. It duplicates the condition in the loop. Maybe move it in the loop and change to `if (next == len)`?

mpaolini committed:
Mimimize the conditionals and the numer of ops inside the
json.loads hot path.

This partially reverts the previous commit as we found out
the culprit wasn't the MOV but the strictness checks

mpaolini committed:
Cleanup indentation

ncoghlan commented on Modules/_json.c:
```diff
@@ -434,16 +434,22 @@ scanstring_unicode(PyObject *pystr, Py_ssize_t end, int strict, Py_ssize_t *next
+            /* Defer the strict error until outside this (hot) loop. */
+            /* See bpo-37587 */
+            if (c <= 0x1f && invalid < 0) {
+                invalid = next;
```
It seems unlikely that an ordering comparison could be faster than a check against zero, so I suspect most or all of the speed-up here is coming from performing the usually-false check before the usually-true one.

What performance numbers do you get with just the reordering of the operands to put the character check first?

My other question would be to ask what happens to the assembly output if you explicitly mark "strict" as const with the original code? (the while loop body is long enough that I doubt compilers will be scanning it to see if "strict" is ever actually modified, so the explicit hint may make a difference)

mpaolini replied:
just flipping around the condition makes it 11% faster than vanilla. moving the goto out of the loop gives you another 1% (maybe just because the loop is smaller)

ncoghlan replied:
In that case, I'd suggest changing the revised check to be:

    if (c <= 0x1f && strict) {
        invalid = next;
        break;
    }

And then not re-checking strict outside the loop - just check for `invalid >= 0`.

(The currently posted version will be much slower when it comes to rejecting bad data, as it will continue a doomed scan, potentially until the end of the file, instead of exiting as soon as it sees an invalid character)

zooba replied:
Being slower when rejecting bad data is okay though, since it's going to result in an exception.

Our "even faster" version defers the actual check until after the loop completes, and we keep the minimum seen `c` throughout the loop (which uses a branchless `cmov` instruction on Intel-like architectures). That way we only check `strict` once.

And I believe you're spot on - the performance improvement from simply switching comes because the first condition being false bypasses a couple of branches, whereas before they were being included, and when `strict==0` it should make no difference. But avoiding the additional branching entirely is even better.

serhiy-storchaka replied:
The last edition looks incorrect to me. Shouldn't `break` be added after `invalid = next`?

I am fine with just swapping `strict` and `c <= 0x1f` if it gives a noticeable effect (10% or like). We can try more subtle changes (with effect ~ 1%) in a separate issue.

mpaolini replied:
we removed the `break` on purpose to make the loop as small as possible, the algorithm is still correct

mpaolini replied:
@ncoghlan done as you suggested, just switched around the `strict` check

mpaolini committed:
Revert some changes not relevant for performance
